pkg/k8shandler/fluentd.go

```diff
@@ -414,6 +414,38 @@ func newFluentdPodSpec(cluster *logging.ClusterLogging, elasticsearchAppName str
 	// Shorten the termination grace period from the default 30 sec to 10 sec.
 	fluentdPodSpec.TerminationGracePeriodSeconds = utils.GetInt64(10)
 
+	// FIXME: The following conditional branch is a refactoring candidate
+	// To address here https://issues.redhat.com/browse/LOG-833
+	//
+	// Why do we need this branch at all?
+	//
+	// The ClusterLogging CR supports three additional use cases without providing
+	// a logStore stanza namely:
+	// Case 1: A collection only stanza to use with secure forward
+	// Case 2: A collection only stanza to use with syslog
+	// Case 3: A collection only stanza to use with ClusterLogFowarder API
+	//
+	// Supporting all three cases implies:
+	// 1. We don't want to add the init container if `cluster.Spec.LogStore == nil`.
+	// 2. We don't want to add the init container if `cluster.Spec.LogStore == nil`
+	//    and a ClusterLogFowarder CR and does not provide a default output (only
+	//    relevant for case 3)
+	//
+	// What is the init container bound to a log store stanza?
+	//
+	// The init container implementation is only relevant for the default ES log store.
+	// It checks for the version of an ES cluster being >=6. In upgrade scenarios
+	// an ES cluster can eventually provide nodes with multiple version (5.x and 6.x)
+	// until all nodes are upgraded. The data model used in 5.x is not backward
+	// compatible to 5.x, thus preventing fluentd from starting avoids data model
+	// inconsistencies during ES upgrades from 5.x to 6.x.
+	//
+	// Why does this work w/o a ClusterLogForwarder CR provided?
+	//
+	// ClusterLogging and ClusterLogFowarder CR reconciliation share the same logic
+	// for collection config generation. If first is provided without the latter a
+	// ClusterLogFowarder with default fields is assumed.
+	// (See ClusterLoggingRequest#getLogForwarder)
 	if pipelineSpec.HasDefaultOutput() && cluster.Spec.LogStore != nil {
```

Comment:
i guess the case i'm still struggling with is "piplinespec that does not have default output, and spec.logstore is not nil"

why does it matter if the pipelinespec has a defaultoutput?  If logstore is not nil, then we have elastic and need the initcontainer, right?  Or is spec.logstore ignored elsewhere(for deploying elastic) when the pipeline has a non-default output?

Reply:
One needs to put the default log store in perspective when a User creates a ClusterLogging with `spec.LogStore` set and a LogForwarding CR together. Regardless of any additional outputs put in the LogFowarding CR, the user will get always one additional output namely the default logstore from `spec.LogStore`. This automatic addition is there to ensure that a running logging stack and especially the logStore is not destroyed by creating a LogForwarding CR on the first usage. The user can opt-out later from using the default logstore.

Reply:
i'm still not seeing what you'd do any differently if the LogForwarding CR(pipelinespec?) does not have a defaultoutput, but cluster.Spec.LogStore is set.

seems like you would still need to add the initcontainer because you're still deploying elastic.

Reply:
Even if the user does not provide a default output in the LogForwarding CR, the `spec.LogStore` is inherited automatically for the user so in turn we have a default output.

However, if the user sets [DisableDefaultForwarding](https://github.com/openshift/cluster-logging-operator/blob/release-4.5/pkg/apis/logging/v1alpha1/forwarding_types.go#L39), this will be translated as an opt-out and anything defined in `spec.LogStore` is not taken into account.

Reply:
> However, if the user sets DisableDefaultForwarding, this will be translated as an opt-out and anything defined in spec.LogStore is not taken into account.

got it, this is the piece i was missing i think.